File: rust/examples/build_and_send_sell.rs
//! Build, sign, and submit a sell transaction.
//!
//! Flow:
//! 1. Request an unsigned sell transaction from the Exit API.
//! 2. Sign it with a local Solana keypair file.
//! 3. Submit it to either a standard RPC endpoint or Helius Sender.
//!
//! Before running:
//! - Replace all placeholders, including `keypair_path`.
//! - Set `amount_tokens_atomic` in the mint's smallest unit.
//! - If `send_target` is `rpc`, set `rpc_url` to a valid endpoint.

use std::error::Error;

use lasersell_sdk::exit_api::{BuildSellTxRequest, ExitApiClient, SellOutput};
use lasersell_sdk::tx::{send_via_helius_sender, send_via_rpc, sign_unsigned_tx};
use secrecy::SecretString;
use solana_sdk::signature::read_keypair_file;

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let api_key = "REPLACE_WITH_API_KEY".to_string();
    let mint = "REPLACE_WITH_MINT".to_string();
    let user_pubkey = "REPLACE_WITH_WALLET_PUBKEY".to_string();
    let keypair_path = "REPLACE_WITH_KEYPAIR_PATH".to_string();
    // "rpc" uses a standard Solana RPC endpoint; "helius_sender" uses Helius Sender.
    let send_target = "rpc".to_string();
    let rpc_url = "REPLACE_WITH_RPC_URL".to_string();

    // Example: for a 6-decimal mint, 1 whole token = 1_000_000 atomic units.
    let amount_tokens_atomic = 1_000_000_u64;
    let slippage_bps = Some(2_000_u16); // 2,000 bps = 20%
    let keypair = read_keypair_file(keypair_path)?;

    let exit_api = ExitApiClient::with_api_key(SecretString::new(api_key))?;
    let http = reqwest::Client::builder().no_proxy().build()?;

    let request = BuildSellTxRequest {
        mint,
        user_pubkey,
        amount_tokens: amount_tokens_atomic,
        slippage_bps,
        mode: None,
        output: Some(SellOutput::Sol),
        referral_id: None,
        market_context: None,
    };

    let unsigned_tx_b64 = exit_api.build_sell_tx_b64(&request).await?;
    let signed_tx = sign_unsigned_tx(&unsigned_tx_b64, &keypair)?;

    let signature = match send_target.as_str() {
        "helius_sender" => send_via_helius_sender(&http, &signed_tx).await?,
        "rpc" => send_via_rpc(&http, &rpc_url, &signed_tx).await?,
        other => {
            return Err(
                format!("send_target must be `rpc` or `helius_sender` (got `{other}`)").into(),
            )
        }
    };

    println!("signature={signature}");
    Ok(())
}

File: python/examples/build_and_send_sell.py
"""Build, sign, and submit a sell transaction.

Flow:
1. Build unsigned sell tx via Exit API.
2. Sign tx using local keypair.
3. Submit tx via standard RPC or Helius Sender.

Before running:
- Replace all placeholders, including keypair path and API key.
- Set `send_target` to `"rpc"` or `"helius_sender"`.
- If using `"rpc"`, set `rpc_url` to a valid Solana RPC endpoint.
"""

from __future__ import annotations

import asyncio
import json
from pathlib import Path

from solders.keypair import Keypair

from lasersell_sdk.exit_api import BuildSellTxRequest, ExitApiClient, SellOutput
from lasersell_sdk.tx import send_via_helius_sender, send_via_rpc, sign_unsigned_tx


async def main() -> None:
    api_key = "REPLACE_WITH_API_KEY"
    keypair_path = Path("REPLACE_WITH_KEYPAIR_PATH")
    send_target = "rpc"  # "rpc" or "helius_sender"
    rpc_url = "REPLACE_WITH_RPC_URL"

    keypair = read_keypair_file(keypair_path)
    exit_api = ExitApiClient.with_api_key(api_key)

    request = BuildSellTxRequest(
        mint="REPLACE_WITH_MINT",
        user_pubkey="REPLACE_WITH_WALLET_PUBKEY",
        amount_tokens=1_000_000,
        slippage_bps=2_000,
        output=SellOutput.SOL,
    )

    unsigned_tx_b64 = await exit_api.build_sell_tx_b64(request)
    signed_tx = sign_unsigned_tx(unsigned_tx_b64, keypair)

    if send_target == "helius_sender":
        signature = await send_via_helius_sender(signed_tx)
    elif send_target == "rpc":
        signature = await send_via_rpc(rpc_url, signed_tx)
    else:
        raise ValueError(f"send_target must be rpc or helius_sender (got {send_target!r})")

    print(f"signature={signature}")


def read_keypair_file(path: Path) -> Keypair:
    raw = path.read_text(encoding="utf-8")
    parsed = json.loads(raw)
    if not isinstance(parsed, list):
        raise ValueError("keypair file must be an array of numbers")

    secret = bytes(int(value) for value in parsed)
    return Keypair.from_bytes(secret)


if __name__ == "__main__":
    asyncio.run(main())
